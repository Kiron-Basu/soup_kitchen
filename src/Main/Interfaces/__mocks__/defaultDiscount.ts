import { IDiscount } from "../IDiscount";

export const defaultDiscount: IDiscount = {
  type: "Bread discount",
  discount: 100,
};
