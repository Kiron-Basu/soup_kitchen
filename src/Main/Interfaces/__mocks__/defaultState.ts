export const defaultState = {
  bread: {
    price: 110,
    quantity: 0,
    subtotal: 0,
  },
  butter: {
    price: 120,
    quantity: 0,
    subtotal: 0,
  },
  cheese: {
    price: 90,
    quantity: 0,
    subtotal: 0,
  },
  milk: {
    price: 50,
    quantity: 0,
    subtotal: 0,
  },
  soup: {
    price: 60,
    quantity: 0,
    subtotal: 0,
  },
};
