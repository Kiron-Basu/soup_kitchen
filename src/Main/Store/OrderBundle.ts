import { FoodItem } from "../Enums/FoodItems";

// Action Types
export enum ActionTypes {
  ADD_ITEM = "ORDER_ADD_ITEM",
  REMOVE_ITEM = "ORDER_REMOVE_ITEM",
  ADD_TO_CART = "ADD_TO_CART",
}

//Action Interfaces
export interface IAddItemAction {
  type: ActionTypes.ADD_ITEM;
  item: FoodItem;
}

export interface IRemoveItemAction {
  type: ActionTypes.REMOVE_ITEM;
  item: FoodItem;
}

export interface IAddToCartAction {
  type: ActionTypes.ADD_TO_CART;
  item: FoodItem;
  quantity: number;
}

// Consolidate Action Interfaces
type Actions = IAddItemAction | IRemoveItemAction | IAddToCartAction;

// Action Creators
export const actionCreators = {
  addItem(item: FoodItem): IAddItemAction {
    return {
      type: ActionTypes.ADD_ITEM,
      item: item,
    };
  },

  removeItem(item: FoodItem): IRemoveItemAction {
    return {
      type: ActionTypes.REMOVE_ITEM,
      item: item,
    };
  },

  addToCart(item: FoodItem, quantity: number): IAddToCartAction {
    return {
      type: ActionTypes.ADD_TO_CART,
      item: item,
      quantity: quantity,
    };
  },
};
export interface IOrderItem {
  price: number;
  quantity: number;
  subtotal: number;
}
// Shape of store
export interface IOrderState {
  bread: IOrderItem;
  cheese: IOrderItem;
  milk: IOrderItem;
  butter: IOrderItem;
  soup: IOrderItem;
}

// Default store
export function getInitialOrderState(): Readonly<IOrderState> {
  return {
    bread: { price: 110, quantity: 0, subtotal: 0 },
    cheese: { price: 90, quantity: 0, subtotal: 0 },
    milk: { price: 50, quantity: 0, subtotal: 0 },
    butter: { price: 120, quantity: 0, subtotal: 0 },
    soup: { price: 60, quantity: 0, subtotal: 0 },
  };
}

// Subreducers

// function processAddItem(
//   state: IOrderState,
//   action: IAddItemAction
// ): IOrderState {
//   return {
//     ...state,
//     [action.item]: state[action.item] + 1, //check if there's a better way to do
//   };
// }

// function processRemoveItem(
//   state: IOrderState,
//   action: IRemoveItemAction
// ): IOrderState {
//   return {
//     ...state,
//     [action.item]: state[action.item] - 1, //check if there's a better way to do
//   };
// }

function processAddToCart(
  state: IOrderState,
  action: IAddToCartAction
): IOrderState {
  return {
    ...state,
    [action.item]: {
      ...state[action.item],
      quantity: action.quantity,
      subtotal: action.quantity * state[action.item].price,
    },
  };
}

// Reducer
export default function orderReducer(
  state: IOrderState = getInitialOrderState(),
  action: Actions
): IOrderState {
  console.log("@@@ in reducer", action);

  switch (action.type) {
    // case ActionTypes.ADD_ITEM:
    //   return processAddItem(state, action);
    // case ActionTypes.REMOVE_ITEM:
    //   return processRemoveItem(state, action);
    case ActionTypes.ADD_TO_CART:
      return processAddToCart(state, action);
    default:
      return state;
  }
}
