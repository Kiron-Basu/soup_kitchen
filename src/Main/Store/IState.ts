import { IOrderState } from "../Store/OrderBundle";

export interface IState {
  orderState: IOrderState;
}
