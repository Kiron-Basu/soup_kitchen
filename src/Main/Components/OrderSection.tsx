import React, { RefObject } from "react";
import { connect } from "react-redux";

import { IState } from "../Store/IState";
import { 
  actionCreators,
  IOrderState 
} from "../Store/OrderBundle";
import { OrderItem } from "../Components/OrderItem";
import { FoodItem } from "../Enums/FoodItems";
import {
  getDiscounts,
  totalBeforeDiscounts,
} from "../Utilities/InvoiceUtilities";
import { IDiscount } from "../Interfaces/IDiscount";
import { OrderSummary } from "./OrderSummary";
import { dispatch } from "../../Store";

interface IStoreProps {
  order: IOrderState;
}

interface IRawProps {
  foodLabels: ReadonlyArray<string>;
  className?: string;
}

export type IOrderSectionProps = IStoreProps & IRawProps;
export const OrderSection: React.FC<IOrderSectionProps> = (
  props
): JSX.Element => {
  const { order, foodLabels } = props;
  const preDiscountTotal: number = totalBeforeDiscounts(order);
  const discounts: ReadonlyArray<IDiscount> = getDiscounts(order);

  const handleClearItem = (item: FoodItem) => {
    dispatch(actionCreators.clearItem(item));
  };
  const ref: RefObject<SVGSVGElement> = React.createRef();
  
  return (
    <div className={props.className}>
      <h2 className={`${props.className}__title`}>Order Summary</h2>
      {foodLabels.map((label) => {
        const quantity: number = order[label as keyof typeof FoodItem].quantity;
        const subtotal: number = order[label as keyof typeof FoodItem].subtotal;
        if (quantity > 0) {
          return (
            <OrderItem
              ref={ref}
              key={label}
              foodItem={label}
              quantity={quantity}
              subtotal={subtotal}
              handleClearItem={handleClearItem}
            />
          );
        }
        return null;
      })}
      <OrderSummary preDiscountTotal={preDiscountTotal} discounts={discounts} />
    </div>
  );
};

function mapStateToProps(state: Readonly<IState>): IStoreProps {
  return { order: state.orderState };
}

export default connect<IStoreProps, {}, {}, IState>(
  mapStateToProps,
  null,
  null,
  { forwardRef: true }
)(OrderSection);
