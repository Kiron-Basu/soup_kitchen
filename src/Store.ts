import { createStore, Reducer, combineReducers, Store } from "redux";
import orderReducer, { getInitialOrderState } from "./Main/Store/OrderBundle";
import { IState } from "./Main/Store/IState";

export let store: Readonly<Store<IState>>;

export function dispatch(action: { type: string }): void {
  if (store !== undefined) {
    store.dispatch(action);
  }
}

function getReducers(): {} {
  return {
    orderState: orderReducer,
  };
}

export default function configureStore(initialState = getInitialOrderState) {
  const reducer: Reducer<{}> = combineReducers(getReducers());
  return createStore(reducer, initialState()) as Store<IState>;
}

export function setStore(setAsStore: Readonly<Store<IState>>): void {
  store = setAsStore;
}
