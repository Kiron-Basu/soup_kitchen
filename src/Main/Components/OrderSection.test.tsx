import React from "react";
import { 
  ShallowWrapper,
  shallow 
} from "enzyme";

import * as Store from "../../Store";
import { OrderSection } from "./OrderSection";
import { defaultState } from "../Interfaces/__mocks__/defaultState";

describe("OrderSection tests", () => {
  const dispatchMock: jest.SpyInstance<void> = jest
    .spyOn(Store, "dispatch")
    .mockImplementation();

  const foodLabels: ReadonlyArray<string> = Object.keys(defaultState);

  test("Renders OrderSection correctly", () => {
    const component: ShallowWrapper = shallow(
      <OrderSection foodLabels={foodLabels} order={defaultState} />
    );

    expect(component).toMatchSnapshot();
  });
});
