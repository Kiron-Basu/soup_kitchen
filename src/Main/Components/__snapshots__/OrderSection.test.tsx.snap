// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderSection tests Renders OrderSection correctly 1`] = `
<div>
  <h2
    className="undefined__title"
  >
    Order Summary
  </h2>
  <OrderSummary
    discounts={
      Array [
        Object {
          "discount": 0,
          "type": "butter discount",
        },
        Object {
          "discount": 0,
          "type": "cheese discount",
        },
        Object {
          "discount": 0,
          "type": "bread discount",
        },
      ]
    }
    preDiscountTotal={0}
  />
</div>
`;
